src/Relations/MorphToMany.php

```diff
@@ -0,0 +1,461 @@
+        }
+
+        if ($touch) {
+            $this->touchIfTouching();
+        }
+    }
+
+    /** @inheritdoc */
+    public function detach($ids = [], $touch = true)
+    {
+        if ($ids instanceof Model) {
+            $ids = $this->parseIds($ids);
+        }
+
+        $query = $this->newRelatedQuery();
+
+        // If associated IDs were passed to the method we will only delete those
+        // associations, otherwise all the association ties will be broken.
+        // We'll return the numbers of affected rows when we do the deletes.
+        $ids = (array) $ids;
+
+        // Detach all ids from the parent model.
+        if ($this->getInverse()) {
+            // Remove the relation from the parent.
+            $data = [];
+            foreach ($ids as $item) {
+                $data = array_merge($data, [
+                    [
+                        $this->relatedPivotKey => $item,
+                        $this->morphType       => $this->related->getMorphClass(),
+                    ],
+                ]);
+            }
+
+            $this->parent->pull($this->table, $data);
+
+            // Prepare the query to select all related objects.
+            if (count($ids) > 0) {
+                $query->whereIn($this->relatedKey, $ids);
+            }
+
+            // Remove the relation from the related.
+            $query->pull($this->foreignPivotKey, $this->parent->{$this->parentKey});
+        } else {
+            // Remove the relation from the parent.
+            $this->parent->pull($this->relatedPivotKey, $ids);
+
+            // Prepare the query to select all related objects.
+            if (count($ids) > 0) {
+                $query->whereIn($this->relatedKey, $ids);
+            }
+
+            // Remove the relation to the related.
+            $query->pull($this->table, [
+                [
+                    $this->foreignPivotKey => $this->parent->{$this->parentKey},
+                    $this->morphType => $this->parent->getMorphClass(),
+                ],
+            ]);
+        }
+
+        if ($touch) {
+            $this->touchIfTouching();
+        }
+
+        return count($ids);
+    }
+
+    /** @inheritdoc */
+    protected function buildDictionary(Collection $results)
+    {
+        $foreign = $this->foreignPivotKey;
+
+        // First we will build a dictionary of child models keyed by the foreign key
+        // of the relation so that we will easily and quickly match them to their
+        // parents without having a possibly slow inner loops for every models.
+        $dictionary = [];
+
+        foreach ($results as $result) {
+            if ($this->getInverse()) {
+                foreach ($result->$foreign as $item) {
+                    $dictionary[$item][] = $result;
+                }
+            } else {
+                // Collect $foreign value from pivot column of result model
+                $items = $this->extractIds($result->{$this->table} ?? [], $foreign);
```

Comment:
Ids extracted from the results pivot column in `MorphToMany` mode to build the dictionary.